src/main/java/net/fabricmc/loader/impl/ObjectShareImpl.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2016 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.loader.impl;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.BiConsumer;
+
+import net.fabricmc.loader.api.ObjectShare;
+
+final class ObjectShareImpl implements ObjectShare {
+	private final Map<String, Object> values = new HashMap<>();
+	private final Map<String, List<BiConsumer<String, Object>>> pending = new HashMap<>();
+
+	@Override
+	public synchronized Object get(String key) {
+		validateKey(key);
+
+		return values.get(key);
+	}
+
+	@Override
+	public synchronized Object put(String key, Object value) {
+		validateKey(key);
+		Objects.requireNonNull(value, "null value");
+
+		Object prev = values.put(key, value);
+
+		if (prev == null) {
+			invokePending(key, value);
+		}
+
+		return prev;
+	}
+
+	@Override
+	public synchronized Object putIfAbsent(String key, Object value) {
+		validateKey(key);
+		Objects.requireNonNull(value, "null value");
+
+		Object prev = values.putIfAbsent(key, value);
+
+		if (prev == null) {
+			invokePending(key, value);
+		}
+
+		return prev;
+	}
+
+	@Override
+	public synchronized Object remove(String key) {
+		validateKey(key);
+
+		return values.remove(key);
+	}
+
+	@Override
+	public synchronized void whenAvailable(String key, BiConsumer<String, Object> consumer) {
+		validateKey(key);
+
+		Object value = get(key);
+
+		if (value != null) {
+			consumer.accept(key, value);
```

Comment:
This and other calls look like a bad idea (tm).

Synchronizing on an object and invoking arbitrary code is just asking for problems (dead locks).

You should rework this so the call outs are done outside the synchronized blocks.

Reply:
Or you could just change the code to use thread safe datastructures and remove the synchronized blocks?

Reply:
How so? there is no way this can deadlock since synchronized supports re-entrant accesses just fine.

A thread safe data structure is insufficient for this kind of application, the interaction between `values` and `pending` needs to be tightly controlled as well.

Reply:
thread1 does:
````
os1.whenAvailable(x -> os2.whenAvailable(...));
````
at the same time as thread2 does:
````
os2.whenAvailable(x -> os1.whenAvailable(...));
````
When things go bad, thread1 holds a lock on os1 and wants os2 while thread2 holds a lock on os2 and wants os1
This is a deadlock.

If you say there is only ever one ObjectShare instance, replace os2 with some other synchronized object doing the same (anti-)pattern.

Reply:
> If you say there is only ever one ObjectShare instance, replace os2 with some other synchronized object doing the same (anti-)pattern.

For such cases, IMO it should really be users' resposibilities to `synchronize` on both objects before doing anything with them, and they should understand that these objects may be synchronized in arbitrary order instead of acquiring one before another etc.

Reply:
Can we use a read-write lock that only locks the write lock for `put` `putIfAbsent` and `remove`? I guess this may be faster than synchronization.

If we do, we probably can call `invokePending` after releasing the read lock so they can modify this share with callbacks.

Reply:
> For such cases, IMO it should really be users' resposibilities to synchronize on both objects

* The relevant objects might be implementation details and therefore inaccessible, this is especially true when this feature is about loose coupling of mods.
* The very nature of whenAvailable() means the user is not in control of the entry point. And the one who is, has no idea what other objects maybe involved.
* Typically, the use cases will be more complicated than the simple example I used above, it might even manifest as chaining of whenAvailable()s across multiple mods.

Reply:
So a read write lock won't work?

Reply:
>So a read write lock won't work?

Attempting a read lock will still deadlock if somebody holds a write lock.
And just read locks won't emulate synchronized.

Off topic: The standard java RentrantReadWriteLock has the annoying "feature" that you can't upgrade a read lock to a write lock (even when you are the only one holding a read lock):
https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/concurrent/locks/ReentrantReadWriteLock.html

Reply:
> Attempting a read lock will still deadlock if somebody holds a write lock.
And vice-versa.

Reply:
Don't think the methods that can acquire the write locks accept callbacks (none of `put` `putIfAbsent` and `remove` executes a callback), so your concern is pretty much extraneous.

Reply:
> Don't think the methods that can acquire the write locks accept callbacks 

They may not accept callbacks, but they do invoke callbacks under lock which can start the chain.

Reply:
This untested  (not even compiled) diff shows what I originally suggested.

It simply invokes the callbacks outside the synchronized blocks while keeping all the data structure logic and locking the same.
````
40c40
< 	public synchronized Object put(String key, Object value) {
---
> 	public Object put(String key, Object value) {
43,48c43,55
< 
< 		Object prev = values.put(key, value);
< 
< 		if (prev == null) {
< 			invokePending(key, value);
< 		}
---
>  
>            Object prev = null;
>            List<BiConsumer<String, Object>> pendingEntries = null;
>            synchronized (this) {
>                prev = values.put(key, value);
>                if (prev == null) {
>                    pendingEntries = pending.remove(key);
>                }
>            }
>  
>            if (prev == null) { 
>                invokePending(key, value, pendingEntries);
>            }
54c61
< 	public synchronized Object putIfAbsent(String key, Object value) {
---
> 	public Object putIfAbsent(String key, Object value) {
58,62c65,76
< 		Object prev = values.putIfAbsent(key, value);
< 
< 		if (prev == null) {
< 			invokePending(key, value);
< 		}
---
>            Object prev = null;
>            List<BiConsumer<String, Object>> pendingEntries = null;
>            synchronized (this) {
>                prev = values.putIfAbsent(key, value);
>                if (prev == null) {
>                    pendingEntries = pending.remove(key);
>                }
>            }
> 
>            if (prev == null) { 
>                invokePending(key, value, pendingEntries);
>            }
75c89
< 	public synchronized void whenAvailable(String key, BiConsumer<String, Object> consumer) {
---
> 	public void whenAvailable(String key, BiConsumer<String, Object> consumer) {
78c92,98
< 		Object value = get(key);
---
>            Object value = null;
>            synchronized (this) {
>                value = get(key);
>                if (value == null) {
>                    pending.computeIfAbsent(key, ignore -> new ArrayList<>()).add(consumer);
>                }
>            }
82,84c102
< 		} else {
< 			pending.computeIfAbsent(key, ignore -> new ArrayList<>()).add(consumer);
< 		}
---
>            }
94,96c112
< 	private void invokePending(String key, Object value) {
< 		List<BiConsumer<String, Object>> pendingEntries = pending.remove(key);
< 
---
> 	private void invokePending(String key, Object value, List<BiConsumer<String, Object>> pendingEntries) {
````

Reply:
You are describing something simple in a complex way. You should just say to call consumers outside of the locks.

Reply:
> You are describing something simple in a complex way. You should just say to call consumers outside of the locks.

That's what I originally said before the long argument. :-)
````
You should rework this so the call outs are done outside the synchronized blocks.
````

Reply:
I changed it since it's trivial to handle nicer in this case, the two thread interaction wouldn't really qualify as a bug on our side though.